internal/worker/dynamic_params.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2017-2021 Uber Technologies Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package worker
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/marusama/semaphore/v2"
+)
+
+var _ Permit = (*permit)(nil)
+
+// Synchronization contains synchronization primitives for dynamic configuration.
+type DynamicParams struct {
+	PollerPermit Permit // controls concurrency of pollers
+	TaskPermit   Permit // controlls concurrency of task processings
+}
+
+// Permit is an adaptive
+type Permit interface {
+	Acquire(context.Context, int) error
+	AcquireChan(context.Context, *sync.WaitGroup) <-chan struct{}
+	Quota() int
+	SetQuota(int)
+	Count() int
+	Release(count int)
+}
+
+type permit struct {
+	sem semaphore.Semaphore
+}
+
+// NewPermit creates a dynamic permit that's resizable
+func NewPermit(initCount int) Permit {
```

Comment:
let's add unit tests  for permit implementation

Reply:
added